### ISSUE TITLE
EPMoE output is padding-sensitive under bf16 (breaks spec-greedy == nospec-greedy on MoE targets)

### ISSUE THREAD

**Issue by zorrofox** [sgl-project/sglang-jax#1090]
### Symptom

On MiMo-V2.5-Pro (384-expert MoE, `--moe-backend epmoe --ep-size 64`, v6e-64), greedy spec-decode output diverges from nospec greedy:

| prompt | nospec (padded `ids.shape=(64,)`) | spec (unpadded `ids.shape=(5,)`) |
|---|---|---|
| `"The capital of France is"` | `[264] " a"` | `[825] " one"` |

After forcing spec prefill through `get_model_worker_batch` (padded, same as nospec):

| | first token | divergence point |
|---|---|---|
| spec (padded prefill) | `[264] " a"` ✓ | token ~30 (verify still unpadded) |

i.e. **same 5 real tokens, different logits depending on whether 59 padding tokens are present**.

### Root cause

`EPMoE._permute` (`layers/moe.py:705-712`):

```python
flatten_selected_experts = jnp.ravel(top_k_indices)          # includes padding tokens' expert ids
sorted_selected_experts = jnp.argsort(flatten_selected_experts, stable=True)
token_indices = sorted_selected_experts // self.num_experts_per_tok
group_sizes = jnp.bincount(flatten_selected_experts, length=self.num_experts)
```

Padding tokens (`embed_tokens[0]`) go through the router and get real expert assignments. After the stable argsort, a real token's row index inside its expert group depends on how many padding tokens picked a `<=` expert id — so the same real token lands in a different GMM tile under different padding. Under bf16 accumulation that's enough to flip argmax when top-1/top-2 logits are close.

`_unpermute` correctly scatters padding outputs back to padding positions (so this is **not** a logical contamination), and the internal 128-align pad at `_gmm_compute:610-613` is trimmed at `_unpermute:728`. The remaining effect is purely **numeric ordering** inside GMM.

### Why this matters for spec decode

- nospec path: `get_model_worker_batch` pads tokens/bs to precompile buckets
- spec path: `get_spec_model_worker_batch` does **not** pad (`eagle_worker.py:60` `# FIXME(pc) add padding logic here`)
- spec verify forwards `bs × num_draft_tokens` real tokens; nospec decode forwards `padded_bs × 1`

So spec target-forward and nospec target-forward see different `total_tokens` at every step → different GMM row layout → spec greedy ≠ nospec greedy on MoE+EP targets. Dense targets (P1-0 Llama-EAGLE3) don't hit this.

### Repro

v6e-64, `epmoe ep=64 dp=1`, `disable-precompile`, MiMo-V2.5-Pro:
1. nospec `max_new_tokens=1` → `[264]`
2. spec (NEXTN, `get_spec_model_worker_batch` unpadded prefill) `max_new_tokens=1` → `[825]`
3. spec with prefill forced through `get_model_worker_batch` (padded) → `[264]` ✓

### Possible fixes

- **A (router-side mask)**: before `_permute`, set `topk_ids[real_num_tokens:] = num_experts` (out-of-range, drops out of `bincount(length=num_experts)`) and `topk_weights[real_num_tokens:] = 0`. Real tokens' sort positions become padding-invariant. Needs `real_num_tokens` plumbed into the MoE layer.
- **B (spec-side padding)**: make `get_spec_model_worker_batch` pad to the same bs/token buckets as nospec. Helps prefill (verified) but verify (`bs × draft_tokens`) vs nospec decode (`padded_bs × 1`) are still different shapes.
- **C (fp32 router/accum)**: doesn't address row-reordering, only narrows the diff.

A is the clean fix but touches the MoE layer API. B is what P1-4 currently does for prefill only.

### Scope

Tracking separately from #1053 P1-4 since it affects any MoE+EP target under spec decode, not just multi-layer MTP.

**Comment by pengchengneo** [comment]
Adding two clarifications:

1. **FusedMoE is likely affected too.** `fused_moe/v1/kernel.py:3172-3175` already handles out-of-range expert ids via a sentinel bucket (`safe_ids = where(valid, ids, num_experts)` + `bincount(length=num_experts+1)[:num_experts]`), so option A on the FusedMoE path doesn't need a kernel API change — only marking padding tokens' `topk_ids` as invalid before the call. Whether the actual numerical drift on FusedMoE is as severe as EPMoE under bf16 needs an A/B run to confirm — I'll verify on v6e-64.

2. **Option A does change nospec numerics.** Logical correctness is preserved (padding outputs are discarded by `_unpermute` either way), but real tokens' GMM row positions shift when padding tokens no longer occupy slots → real tokens' bf16 accumulation order changes → nospec greedy output may produce different tokens than today's nospec for the same input. The upside is the result becomes truly padding-invariant. We should run an accuracy regression (GSM8K/MMLU) before adopting A as default, and flag it as a behavior change to downstream consumers.


**The more likely root cause is per-tile dynamic activation quantization in megablox gmm (maybe_quantize_lhs=True):** the LHS scale is taken over all rows in a tile, so T1's quant scale shifts based on which neighbors (real / padding) share its tile.

Quick ablation to confirm: rerun spec vs nospec with act_q_dtype=None. If outputs become bit-identical, activation quant is the root cause. Will run on v6e-64 and report.

**Comment by pengchengneo** [comment]
The "row index → GMM tile → bf16 accumulation" mechanism is falsified in isolation: standard matmul and the megablox GMM kernel both produce bit-equal output for the same real token at different row positions within an expert group. 

So the root-cause analysis as stated doesn't hold.

The observed spec≠nospec divergence is still real — but the actual mechanism is somewhere else (EP collective ordering / JIT kernel variant / fused_ep_moe path / something I haven't traced yet). Before we land option A, I'd suggest a controlled repro that strips one component at a time to localize the actual cause.

```
"""Minimal repro: does megablox GMM kernel produce bit-different output
when the same real token is at different row positions inside an expert group?

Setup:
- Use the same gmm function that EPMoE._gmm_compute calls.
- Construct two scenarios that route the SAME real tokens to the SAME
  expert, but with different padding tokens preceding them.
- Compare bit-by-bit.

Hypothesis from #1090: real tokens land in different "GMM tile" → bf16
accumulation order changes → bits differ.
"""
import sys
sys.path.insert(0, "/tmp/sglang-jax/python")

import jax
import jax.numpy as jnp
import numpy as np

from sgl_jax.srt.kernels.gmm.megablox_gmm_backend import gmm

print(f"jax devices: {jax.devices()}")
print(f"jax version: {jax.__version__}\n")

NUM_EXPERTS = 8
K = 4096       # hidden_dim
N = 1408       # intermediate_dim per expert
DTYPE = jnp.bfloat16

# Random shared weight: rhs (num_groups, k, n)
key = jax.random.PRNGKey(42)
k_w, k_real, k_pad = jax.random.split(key, 3)
RHS = jax.random.normal(k_w, (NUM_EXPERTS, K, N), dtype=DTYPE)

# Three "real" tokens we want to track
REAL = jax.random.normal(k_real, (3, K), dtype=DTYPE)


def run_gmm(lhs, group_sizes):
    """Wrapped to keep kwargs minimal and match EPMoE call site."""
    out = gmm(
        lhs=lhs,
        rhs=RHS,
        group_sizes=group_sizes.astype(jnp.int32),
        preferred_element_type=DTYPE,
        acc_dtype=jnp.float32,
        maybe_quantize_lhs=False,
        zero_initialize=False,
    )
    return out


# ---- Scenario A: 3 real tokens all routed to expert 0, no padding ----
lhs_A = REAL  # shape (3, K)
gs_A = jnp.array([3, 0, 0, 0, 0, 0, 0, 0])
out_A = run_gmm(lhs_A, gs_A)
out_A_real = np.asarray(out_A)[:3]   # rows 0..2 are the real tokens

# ---- Scenario B: 5 padding tokens (random nonzero) THEN 3 real tokens, all expert 0 ----
PAD = jax.random.normal(k_pad, (5, K), dtype=DTYPE)
lhs_B = jnp.concatenate([PAD, REAL], axis=0)   # (8, K)
gs_B = jnp.array([8, 0, 0, 0, 0, 0, 0, 0])
out_B = run_gmm(lhs_B, gs_B)
out_B_real = np.asarray(out_B)[5:8]   # real tokens are now at rows 5..7

# ---- Scenario C: real tokens at start, padding AFTER (zero pad) ----
ZERO_PAD = jnp.zeros((5, K), dtype=DTYPE)
lhs_C = jnp.concatenate([REAL, ZERO_PAD], axis=0)
gs_C = jnp.array([8, 0, 0, 0, 0, 0, 0, 0])
out_C = run_gmm(lhs_C, gs_C)
out_C_real = np.asarray(out_C)[:3]

# ---- Scenario D: real tokens spread across 3 different experts ----
# T0 → expert 0, T1 → expert 1, T2 → expert 2
gs_D = jnp.array([1, 1, 1, 0, 0, 0, 0, 0])
out_D = run_gmm(REAL, gs_D)
# Each row uses a different expert, so won't compare bits with A/B/C; just sanity that it runs.

# ---- Scenario E: real tokens at end, padding BEFORE in different group_sizes split
# 5 padding to expert 0, 3 real to expert 1 (different expert! so different W)
gs_E = jnp.array([5, 3, 0, 0, 0, 0, 0, 0])
out_E = run_gmm(jnp.concatenate([PAD, REAL], axis=0), gs_E)
out_E_real = np.asarray(out_E)[5:8]   # real tokens went through expert 1


def cmp(name, ref, other):
    eq = np.array_equal(ref, other)
    print(f"[{name}]  bit-equal vs A: {eq}")
    if not eq:
        diff = (ref.view(np.uint16) != other.view(np.uint16))
        n_diff_rows = diff.any(axis=1).sum()
        n_diff_elems = diff.sum()
        max_abs = np.abs(ref.astype(np.float32) - other.astype(np.float32)).max()
        print(f"  diff rows: {n_diff_rows}/{ref.shape[0]}, "
              f"diff elems: {n_diff_elems}/{ref.size}, "
              f"max |Δ|: {max_abs:.6e}")


print("=== GMM row-position bit-stability test ===\n")
print(f"Real tokens shape: {REAL.shape}, K={K}, N={N}, num_experts={NUM_EXPERTS}\n")

print("Scenario A: real tokens at rows 0..2, group_sizes=[3,0,...] → expert 0")
print("Scenario B: pad at 0..4 + real at 5..7, group_sizes=[8,0,...] → all expert 0")
print("Scenario C: real at 0..2 + zero-pad at 3..7, group_sizes=[8,0,...] → all expert 0\n")

cmp("B vs A — same expert, same real tokens, different row position", out_A_real, out_B_real)
cmp("C vs A — same expert, real tokens up front, zero pad after",     out_A_real, out_C_real)

print("\nNote: D and E use different experts on real tokens, so different W → not bit-comparable.")
print("They just verify the kernel runs across multiple group_sizes shapes:")
print(f"  D group_sizes=[1,1,1,...]  output shape {out_D.shape}, all-finite: {np.all(np.isfinite(out_D))}")
print(f"  E group_sizes=[5,3,0,...]  output shape {out_E.shape}, all-finite: {np.all(np.isfinite(out_E))}")

```

**Comment by zorrofox** [comment]
The repro is conclusive — withdrawn the row-position/accumulation-order claim. Activation-quant per-tile scale is a much better fit (padding tokens shifting the shared LHS scale would explain why the effect tracks total_tokens, not just sort order). Our v6e-64 NEXTN job is still up if it helps for the `act_q_dtype=None` ablation — happy to run a spec-vs-nospec first-token comparison with that toggled. The padded-prefill workaround in #1089 stays orthogonal (it just makes spec/nospec see the same tile composition, whatever the underlying mechanism).

**Comment by pengchengneo** [comment]
hi @zorrofox  
any update about this problem

**Comment by zorrofox** [comment]
No new ablation data yet — the v6e-64 slice was repurposed 5/17-5/18 and is currently running the P1-5b spec+DP iteration. Two notes though:
- The user-visible symptom (spec first token ≠ nospec) is already addressed by the padded-prefill workaround in #1089 (spec prefill now goes through `get_model_worker_batch` so target sees the same tile composition as nospec). That doesn't explain *why* padding matters, but it unblocks spec correctness.
- For the root cause: I can slot an `act_q_dtype=None` spec-vs-nospec first-token comparison between P1-5b rounds (~30min, 2 restarts at dp=1) if you haven't already run it on your side — just confirm and I'll queue it.

**Comment by pengchengneo** [comment]
actually, sgl-jax current moe's implement is act_q_dtype=None, so the quantization reason is not solid @zorrofox

**Comment by zorrofox** [comment]
Understood — that rules out activation-quant. Both hypotheses I had (row-position, then act_q) are dead, so I'll stop guessing. The padded-prefill workaround in #1089 (now merged) still holds as the user-facing fix; happy to help instrument whatever direction you want to try next for the underlying cause (e.g. dump per-layer logit max-diff between padded/unpadded prefill on the v6e-64 job to localize which layer first diverges).

**Comment by pengchengneo** [comment]
confirm our current progress: we have completed the DP compatibility and MIMO multilayer adaptations, right? 

Therefore, Phase 1 is only missing the tests for non-greedy sampling and top-k > 1. Is my understanding correct? @zorrofox

**Comment by zorrofox** [comment]
Mostly — MiMo multilayer (P1-4, #1089) and the GMM tile fix (#1104) are merged. DP compatibility (P1-5b, #1108 draft) is still in progress: dp=2 single-req is verified token-identical to nospec; multi-req at dp>1 is being debugged right now (post-partial-finish unbalanced layout — last `[:real_bs]` callers in `get_eagle_multi_step_metadata`). The CPU shape test (22 cases) and the spec_info global-flat↔DP-padded scatter/gather are in. ETA dp=4 verification this week. So from our side P1-5b is the remaining open item; non-greedy (P1-6) and topk>1 (P1-7) are on your list per the split.
